agent/artifact_uploader.go

```diff
@@ -124,29 +124,36 @@ func (a *ArtifactUploader) Collect(ctx context.Context) ([]*api.Artifact, error)
 		seenPaths:        make(map[string]bool),
 	}
 
-	wctx, cancel := context.WithCancelCause(ctx)
-	for _, globPath := range strings.Split(a.conf.Paths, ArtifactPathDelimiter) {
-		globPath := strings.TrimSpace(globPath)
-		if globPath == "" {
-			continue
-		}
+	filesCh := make(chan string)
 
+	// Create a few workers to process files as they are found.
+	// Because a single glob could match many many files, a fixed number of
+	// workers will avoid slamming the runtime (as could happen with a
+	// goroutine per file).
+	wctx, cancel := context.WithCancelCause(ctx)
+	for i := 0; i < runtime.NumCPU(); i++ {
```

Comment:
I think this can cause problems when the agent is run in a container. `runtime.NumCPU` may not respect the limits set by the container orchestration framework about the number of CPUs. At least if you do this
```suggestion
	for i := 0; i < runtime.GOMAXPROCS(0); i++ {
```
then this will be configurable and users can make their configuration match between the agent and the container orchestration framework.

To test this, I ran this program:

```go
package main

import (
	"fmt"
	"runtime"
)

func main() {
	fmt.Printf(
		"NumCPU=%d GOMAXPROCS=%d\n",
		runtime.NumCPU(),
		runtime.GOMAXPROCS(0),
	)
}
```
in docker:

```shell
❯ docker run --rm -it -v "$PWD:/app" -w /app --env GOMAXPROCS=1 --cpus=1 golang:1.21 go run .
NumCPU=20 GOMAXPROCS=1
❯ docker run --rm -it -v "$PWD:/app" -w /app --cpus=1 golang:1.21 go run .
NumCPU=20 GOMAXPROCS=20
```

There's also this: https://github.com/uber-go/automaxprocs, which I don't think we need to use, though one day a customer may ask us to.

Reply:
Everything else look great!